ably/realtime_presence.go

```diff
@@ -0,0 +1,230 @@
+	"github.com/ably/ably-go/ably/proto"
+)
+
+// RealtimePresence represents a single presence map of a prarticular channel.
+// It allows entering, leaving and updating presence state for the current
+// client or on behalf of other client.
+type RealtimePresence struct {
+	mu      sync.Mutex
+	sync    sync.WaitGroup
+	data    string
+	serial  string
+	subs    *subscriptions
+	channel *RealtimeChannel
+	members map[string]*proto.PresenceMessage
+	stale   map[string]struct{}
+	state   proto.PresenceState
+	insync  bool
+	synced  bool
+}
+
+func newRealtimePresence(channel *RealtimeChannel) *RealtimePresence {
+	return &RealtimePresence{
+		subs:    newSubscriptions(subscriptionPresenceMessages),
+		channel: channel,
+		members: make(map[string]*proto.PresenceMessage),
+	}
+}
+
+func (pres *RealtimePresence) send(msg *proto.PresenceMessage, result bool) (Result, error) {
+	if _, err := pres.channel.attach(false); err != nil {
+		return nil, err
+	}
+	switch state := pres.channel.State(); state {
+	case StateChanDetached, StateChanDetaching, StateChanClosing, StateChanClosed,
+		StateChanFailed:
+		return nil, newError(91001, errors.New(StateText(state)))
+	}
+	protomsg := &proto.ProtocolMessage{
+		Action:   proto.ActionPresence,
+		Channel:  pres.channel.state.channel,
+		Presence: []*proto.PresenceMessage{msg},
+	}
+	return pres.channel.send(protomsg, result)
+}
+
+func (pres *RealtimePresence) syncStart() {
+	if pres.insync {
+		return
+	}
+	pres.sync.Add(1)
+	pres.insync = true
+	pres.stale = make(map[string]struct{}, len(pres.members))
+	for memberKey := range pres.members {
+		pres.stale[memberKey] = struct{}{}
+	}
+}
+
+func (pres *RealtimePresence) syncEnd() {
+	if !pres.insync {
+		return
+	}
+	for memberKey := range pres.stale {
+		delete(pres.members, memberKey)
+	}
+	for memberKey, presence := range pres.members {
+		if presence.State == proto.PresenceAbsent {
+			delete(pres.members, memberKey)
+		}
+	}
+	pres.stale = nil
+	pres.insync = false
+	pres.sync.Done()
+}
+
+func (pres *RealtimePresence) handle(msg *proto.ProtocolMessage, syncSerial string) {
+	for _, presmsg := range msg.Presence {
+		if presmsg.ConnectionId == "" {
+			presmsg.ConnectionId = msg.ConnectionId
+		}
+		if presmsg.Timestamp == 0 {
+			presmsg.Timestamp = msg.Timestamp
+		}
+	}
+	pres.mu.Lock()
+	if syncSerial != "" {
+		pres.serial = syncSerial
+		pres.syncStart()
+	}
+	// Account old and ignored presence messages.
+	old := make(map[int]struct{})
+	// Update presence map / channel's member state.
+	for i, member := range msg.Presence {
+		memberKey := member.ConnectionId + member.ClientID
+		if oldMember, ok := pres.members[memberKey]; ok {
+			if member.Timestamp <= oldMember.Timestamp {
+				old[i] = struct{}{}
+				continue
+			}
+		}
+		switch member.State {
+		case proto.PresenceUpdate:
+			memberCopy := *member
+			member = &memberCopy
+			member.State = proto.PresencePresent
+			fallthrough
+		case proto.PresencePresent:
+			delete(pres.stale, memberKey)
```

Comment:
When sync is in progress, I remove a member from stale list when it's reported to be present. Then when sync is complete, I remove all stale members from the presence map [here](https://github.com/ably/ably-go/commit/3c65c16491d27c1c5b4c661e4b092bd3e9fdf276#diff-68e24634a97440d2db9491b85323cea2R74).

When sync is completed, this line is a nop, since stale list is a nil, and deleting from nil map is a nop. The stale list is created when we start sync [here](https://github.com/ably/ably-go/commit/3c65c16491d27c1c5b4c661e4b092bd3e9fdf276#diff-68e24634a97440d2db9491b85323cea2R64).